### DISCUSSION TITLE
How to send Canned Audio Data over an SCO Connection?

### DISCUSSION THREAD

**Discussion by HarryKrause** [google/bumble#560] [Q&A]
Hello Bumble dev's,

I am trying to edit the `bumble\examples\run_hfp_handsfree.py` file to send HFP audio to a paired (bonded+connected) HFP Gateway (AG) device, but I can't seem to get it to work.
 
Here's what I'm attempting to do:
 
I pair the AG to the HF, connect them, start a WiFi call using Google Meet between the AG and a secondary phone, and then see that the SCO connection is established. I've edited the example code to then read from a .wav file, encode the data as an HCI_SynchronousDataPacket , and send the encoded packet from the HF using Host.send_hci_packet() . But I always get the following warning from the USB transport:
 
```
DEBUG:bumble.host:### HOST -> CONTROLLER: SCO: handle=0x0135, ps=0, data_total_length=8, data=494e464f49534654
WARNING:bumble.transport.usb:unsupported packet type 3
```
 
It then continues to attempt to send SCO packets but I do not see anymore of the above warnings and I never hear anything on the secondary phone (receiver on the WiFi call).
 
```
DEBUG:bumble.host: ### HOST -> CONTROLLER: SCO: handle=0x0135, ps=0, data_total_length=8, data=0d0000004c617666
DEBUG:bumble.host: ### HOST -> CONTROLLER: SCO: handle=0x0135, ps=0, data_total_length=8, data=36302e332e313030
DEBUG:bumble.host: ### HOST -> CONTROLLER: SCO: handle=0x0135, ps=0, data_total_length=8, data=00006461746100a0
DEBUG:bumble.host: ### HOST -> CONTROLLER: SCO: handle=0x0135, ps=0, data_total_length=8, data=4b02000000000000...
```
 
From these logs, I'm unable to determine if (1) I'm encoding the SCO packets incorrectly, (2) I'm using send_hci_packet() incorrectly, or (3) bumble does not support sending SCO packets via HCI.
 
Could someone please let me know if sending SCO packets via HCI is supported or not? If so, then is there an example that I can reference for how to do so? I've been referencing the `run_hfp_gateway.py` example, because it seems to send SCO packets via HCI using a .wav file.
 
My configuration is:
OS: Windows 11

**HFP HF device**
HW: Intel Wireless Bluetooth (built-in to my laptop).
Drivers: WinUSB v6.17600.16385.

**HFP AG device**
HW: Samsung Galaxy A51
OS: Android 13

I've attached a [HFP_HF.log](https://github.com/user-attachments/files/17119548/HFP_HF.log) for additional context as needed. Thanks in advance!

**Comment by barbibulle** [comment]
The reason this doesn't work is that the USB transport implementation here doesn't yet support sending/receiving SCO packets to the controller (the main reason for this is that SCO over HCI is not supported by all controllers, since usually the audio exchanged for HFP gets routed from a separate line from the controller to some audio subsystem on the host, typically i2s).
However, what you're trying to do is a perfectly valid use case, and should be supported. I do have a local branch where I started adding support for that, I will try to finish it up this week and push a PR.
NOTE: other transports may support sending SCO over HCI, like for example going through the Android HCI HAL via a proxy, see https://google.github.io/bumble/extras/android_remote_hci.html for more details on that, if you have an Android phone to use for that purpose.

**Reply by HarryKrause** [reply]
@barbibulle Thanks for the quick response! We will look into the Android HCI HAL work-around as you've linked, but the USB transport support for SCO packets is definitely the path that I would prefer to use (one less phone needed for the test set-up). I will await the PR for your local change and test / review once it's ready to go.

**Reply by barbibulle** [reply]
One more quick note on this: not all Bluetooth controllers support SCO packets over HCI, so even with the support for it in the usb transfer implementation, it isn't guaranteed it will work with your particular hardware. We'll find out when you try it.

**Comment by acieszkiel** [comment]
Hey Gilles, thanks for your help in: https://github.com/google/bumble/issues/875#issuecomment-3806617180

I got another question: As I checked in the code of usb transport: https://github.com/google/bumble/blob/main/bumble/transport/usb.py#L486 sending/receiving SCO packets from the controller is still not supported, is that correct?

Any planes to add that in the near feature?

Thank you in advance for your response.

**Reply by barbibulle** [reply]
I did start a PR with support for SCO over USB, but I have not yet been able to get it to work end to end. Not sure yet if the issue is with the USB dongles I have been using (not all support SCO over HCI/USB), or the libusb layer I'm using (only tested on macOS so far), or something else. I plan on working on it a bit more soon.

**Reply by acieszkiel** [reply]
Hi, thank you for the answer.

I've been trying to add SCO support over USB and I managed to receive SCO packets from my BT dongle (Realtek rtl8761bu) and decode them into an audible format (in was encoded with mSBC in my case). It works both for `usb` and `pyusb` transport. But my problem is with sending SCO back to the device. Although it does not return any errors from USB, when I send SCO packets to the Isochronous Endpoint Out, I can't hear anything on the other side. 

In your tests so far, have you managed to send SCO to the BT controller and actually hear it? Have you maybe used a Realtek BT dongle?
I am wondering if I have to yet enable my Realtek BT dongle so it could receive SCO as some of the devices I've seen needs some vendor specific HCI commands to be sent for SCO enabling but I did not find any reference about that in the rtl8761b doc.

So far I've tried to send to the USB:
 - exactly what I receive from the USB (echo),
 - correct mSBC frames without missing or corrupted frames (I prepared SCO packets encapsulating mSBC frames based on what I received but without any broken ones),
 - I added starting Byte 03 (SCO data) at the beginning of each SCO packet,
 - I disabled ISO_IN link (just to avoid any interference),
 - send SCO packets continuously with a delay: 1ms/10ms/100ms
 - tried different connection like voip, a real call etc.

But so far I failed.

Thanks for any help.

**Reply by barbibulle** [reply]
I have run into the same issue. I can receive audio, but not send SCO audio over USB. HFP works both ways with non-USB controllers, but with USB something isn't working right. I've tried with Realtek and a couple other non-Realtek controllers, I get the same result...
I've pushed by working branch, so you can take a look and give it a try and compare: PR #896
